From 848cb3a988412ac2b56063fe65654224ef17b022 Mon Sep 17 00:00:00 2001
From: ayaka <ayaka@soulik.info>
Date: Mon, 25 Jan 2016 01:02:04 +0800
Subject: [PATCH 11/12] v4l2: Configure encoder output dimension with the same
 values
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

We had a V4L2 encoder that was triggering errors because
the output resolution was always kept unchanged even after
negotiating a different valid input resolution using the input caps.
The patch in this merge request fixes that issue by reusing the
input caps to reconfigure the output resolution in the encoder.

This patch comes from Enrique Ocaña Gonzále.

Signed-off-by: memeka <mihailescu2m@gmail.com>
---
 sys/v4l2/gstv4l2videoenc.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/sys/v4l2/gstv4l2videoenc.c b/sys/v4l2/gstv4l2videoenc.c
index 3e4c4d7..e77f20c 100644
--- a/sys/v4l2/gstv4l2videoenc.c
+++ b/sys/v4l2/gstv4l2videoenc.c
@@ -459,6 +459,17 @@ gst_v4l2_video_enc_handle_frame (GstVideoEncoder * encoder,
   }
 
   if (NULL != outcaps) {
+    /* Filling caps should be done in gst_video_encoder_negotiate_default(),
+     * but we need it to be done here in advance because
+     * gst_v4l2_object_set_format() must be done now and can't wait for
+     * caps negotiation. */
+    GST_DEBUG_OBJECT (self, "Filling src caps with output dimensions %ux%u",
+        self->input_state->info.width, self->input_state->info.height);
+
+    gst_caps_set_simple (outcaps,
+        "width", G_TYPE_INT, self->input_state->info.width,
+        "height", G_TYPE_INT, self->input_state->info.height, NULL);
+
     GstBufferPool *pool = GST_BUFFER_POOL (self->v4l2output->pool);
     gst_v4l2_object_set_format (self->v4l2capture, outcaps);
 
-- 
2.7.4

